src/pyfmi/simulation/assimulo_interface_fmi2.pyx

```diff
@@ -279,48 +279,50 @@ cdef class FMIODE2(cExplicit_Problem):
+            if self._f_nbr == 0:
+                if self._logging:
+                    msg = preface + '</%s>'%(solver_info_tag)
+                    self._model.append_log_message("Model", 6, msg)
+                
+                return np.array([[0.0]])
 
-        A = self._model._get_A(add_diag=True, output_matrix=self._A)
-        if self._A is None:
-            self._A = A
+            A = self._model._get_A(add_diag=True, output_matrix=self._A)
+            if self._A is None:
+                self._A = A
 
-        if self._extra_f_nbr > 0:
-            if hasattr(self._extra_equations, "jac"):
-                if self._sparse_representation:
+            if self._extra_f_nbr > 0:
+                if hasattr(self._extra_equations, "jac"):
+                    if self._sparse_representation:
 
-                    Jac = A.tocoo() # Convert to COOrdinate
-                    A2 = self._extra_equations.jac(y_extra).tocoo()
+                        Jac = A.tocoo() # Convert to COOrdinate
+                        A2 = self._extra_equations.jac(y_extra).tocoo()
 
-                    data = np.append(Jac.data, A2.data)
-                    row  = np.append(Jac.row, A2.row+self._f_nbr)
-                    col  = np.append(Jac.col, A2.col+self._f_nbr)
+                        data = np.append(Jac.data, A2.data)
+                        row  = np.append(Jac.row, A2.row+self._f_nbr)
+                        col  = np.append(Jac.col, A2.col+self._f_nbr)
 
-                    # Convert to compressed sparse column
-                    Jac = sps.coo_matrix((data, (row, col)))
-                    Jac = Jac.tocsc()
+                        # Convert to compressed sparse column
+                        Jac = sps.coo_matrix((data, (row, col)))
+                        Jac = Jac.tocsc()
+                    else:
+                        Jac = np.zeros((self._f_nbr+self._extra_f_nbr,self._f_nbr+self._extra_f_nbr))
+                        Jac[:self._f_nbr,:self._f_nbr] = A if isinstance(A, np.ndarray) else A.toarray()
+                        Jac[self._f_nbr:,self._f_nbr:] = self._extra_equations.jac(y_extra)
                 else:
-                    Jac = np.zeros((self._f_nbr+self._extra_f_nbr,self._f_nbr+self._extra_f_nbr))
-                    Jac[:self._f_nbr,:self._f_nbr] = A if isinstance(A, np.ndarray) else A.toarray()
-                    Jac[self._f_nbr:,self._f_nbr:] = self._extra_equations.jac(y_extra)
+                    raise FMUException("No Jacobian provided for the extra equations")
             else:
-                raise FMUException("No Jacobian provided for the extra equations")
-        else:
-            Jac = A
-            
-        if self._logging:
-            msg = preface + '</%s>'%(solver_info_tag)
-            self._model.append_log_message("Model", 4, msg)
+                Jac = A
+        finally:
+            # make sure to always generate valid XML, even if jacobian eval fails
+            if self._logging:
```

Comment:
I just think we should add a boolean after line 291 such that we can verify we have successfully called `append_log_message` before we try to add closing tag below. I assume `self._update_model(t, y)` can fail or `self._f_nbr != 0` which would in turn not generate the log message, and then we end up here if anything fails and we are back to invalid XML. Proposal:

```
try:
  ...
  if ...
    append_log_message
    requires_closing_tag = True

  finally:
    if requires_closing_tag:
      append_log_message
```

Reply:
Made some changes; please have another look.

Also learned this here:

```
def f():
    try:
        return 1
    finally:
        print("a")

a = f() # a = 1 & prints 
```

Reply:
Thats nice I didnt know about that either, thanks for sharing!